core/test/org/javarosa/xpath/expr/test/XPathPathExprTest.java

```diff
@@ -105,6 +115,29 @@ public void doTests() {
         testEval("/data/places/country[@id ='one']/name", instance, null, "Singapore");
     }
 
+    /**
+     * Some simple xpath expressions with multiple predicates that operate over
+     * nodesets.
+     */
+    private void testNestedMultiplicities() {
+        FormParseInit fpi = new FormParseInit("/test_nested_multiplicities.xml");
+        FormDef fd = fpi.getFormDef();
+        FormEntryModel fem = fpi.getFormEntryModel();
+
+        testEval("/data/bikes/manufacturer/model[@id='pista']/@color",
+                fd.getInstance(), null, "seafoam");
+        testEval("join(' ', /data/bikes/manufacturer[@american='yes']/model[.=1]/@id)",
+                fd.getInstance(), null, "karate-monkey vamoots");
+        testEval("join(' ', /data/bikes/manufacturer[@american='yes'][model=1]/model/@id)",
+                fd.getInstance(), null, "vamoots");
+        // if the model elements don't all have values, then the predicate
+        // doesn't evaluate correctly, and we take that as not passing,
+        // returning empty.
+        // TODO: should throw XPathExpression exception?
+        testEval("join(' ', /data/bikes/manufacturer[@american='no'][model=1]/model/@id)",
```

Comment:
model = 1 makes sense to me in this setting, actually. You're only including nodes into the nodeset for which model has the element value 1 (IE: <model>1</model>) and including the ID if so, so that feels like a reasonable thing to be able to do?